Autoscaler implementation

ramr opened the pull request:
This is step 2 of the autoscaler implementation - building on the types defined by @pweil-  in https://github.com/GoogleCloudPlatform/kubernetes/pull/5492 
and the original autoscaler proposal from @smarterclayton at: #2863

To test, define a replication controller and an autoscaler configuration - example at: https://github.com/ramr/autoscaler-k8s-test/blob/master/test-as.json
and the replication controller gets scaled automatically to the min/max limits set in the autoscaler spec.

Or you can use the test repo at https://github.com/ramr/autoscaler-k8s-test

```
git clone https://github.com/ramr/autoscaler-k8s-test  
cd autoscaler-k8s-test  
make K8S_SOURCE_DIR=~/work/kubernetes test  
```

and the replicas for the sample replication controller `cluster/kubectl.sh  get rc`, should scale upto a limit of 6 replicas.

PTAL - Any comments/feedback greatly appreciated. Thx

Commit:
auto-scaler types

Commit:
updates based on feedback for types, comments, etcd factoring, removal of Enabled field

ramr committed:
Checkpoint in-progress autoscaler work.
   * cleanup after code rebase:
   * add PrepareForCreate method to node and namespace strategies
   * fixes for updates to method signatures
   * remove unused imports
   * add the generated conversion functions for the autoscaler types
   * use new rest.StandardStorage for the registry storage
   * test fixes and added new tests
   * add duration and scale by fields
   * add autoscaler interfaces and default generic implementation
   * add scale reconciler
   * add autoscaler monitoring sources support
   * add autoscaler client methods + fake client support

ramr committed:
Autoscaler work changes:
  * Refactor manager
  * Convert to using a work queue
  * Hook into kubernetes and controller manager code
  * Reorganize plugins
  * Copyright
  * Remove deprecated api code for merge/rebase w/ @pweil- changes
  * Some cleanup and renames
  * Remove merged labels
  * More refactoring
  * Add autoscaler registry pieces
  * Setup autoscale manager
  * Add and fix tests
  * Optimize policy assessment if there's no targets to operate on.
  * Add info log messages.

ramr commented:
Signed the CLA.

ramr commented:
@pweil-  Think the bot requires you to agree as well - since I used the original types from the unmerged PR - https://github.com/GoogleCloudPlatform/kubernetes/pull/5492

pweil- commented:
Confirmed.  I am @pweil- and I endorse this PR (politician smile)

ramr committed:
Change validation routine parameters to new,old form for clarity +
add a new strategy for updating autoscaler status.

ramr committed:
Add autoscalers alias and add autoscalers alias to "all".

ramr commented:
(bump) + any way to check the failing builds?
Shippable says ```The build you're looking for wasn't found.```.

@smarterclayton FYI

smarterclayton commented:
ok to test

smarterclayton commented:
```
 ./hack/verify-description.sh
API file is missing the required field descriptions: ./pkg/api/v1beta3/types.go
API file is missing the required field descriptions: ./pkg/api/v1/types.go
```

ramr committed:
Add missing description fields for the autoscaler types.

ramr commented:
:sos:   can't see the ```Shippable``` errors ``` The build you're looking for wasn't found.```?
Would be nice to know what tests the shippable stuff is running ...

smarterclayton commented:
See the shippable yaml file

> On Jun 16, 2015, at 10:38 PM, ramr <notifications@github.com> wrote:
> 
>  can't see the Shippable errors The build you're looking for wasn't found.?
> Would be nice to know what tests the shippable stuff is running ...
> 
> —
> Reply to this email directly or view it on GitHub.
>

ramr committed:
Add files modified by hack/run-gendocs.sh

ramr committed:
Add files generated by  hack/update-swagger-spec.sh

ramr commented:
k - third time's a charm!!  PTAL - thx.

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -1976,3 +1976,136 @@ type RangeAllocation struct {
+// AutoScaleThreshold is a behavior based on statistics used to drive the auto-scaler in scaling decisions.
+type AutoScaleThreshold struct {
+	// Type is the type of threshold being used, intention or value.
+	Type AutoScaleThresholdType `json:"type,omitempty" description:"auto-scale threshold type; intention or value based"`
```
it wasn't clear to me how value-based works; intention-based is configured in Intentions field, but where is value-based configured?

ramr replied:
Good question - don't think it was anywhere but am not sure if I have the right answer to that. 

The types were as defined in the original autoscaler types PR  ttps://github.com/GoogleCloudPlatform/kubernetes/pull/5492, and when I started to use them, found a couple of missing configuration pieces to evaluate the thresholds - the value and duration in particular, which I added to the ```AutoScaleIntentionThresholdConfig``` struct. Given what the intention thresholds support now, the value based thresholds could simply be folded into the intention based ones (aka just need the intention based thresholds - value based becomes a subset).

Don't know the history/discussions leading to the initial autoscaler types PR, so maybe @pweil-  or @smarterclayton  can shed more light on that.  @pweil- / @smarterclayton ?

pweil- replied:
As the discussion progressed intention based threshold were the way everyone was leaning.  I eventually removed the value based thresholds from the code (https://github.com/GoogleCloudPlatform/kubernetes/pull/5492#discussion_r26480673).  The type is probably unnecessary at this point unless we intend to add value-based configs.

ramr committed:
Rebase code and add fixes for changed code due to rebase.

ramr committed:
Add changed generated swagger spec.

ramr commented:
Ok - this has been around for a while, so had to do some babysitting - rebased this to the latest code. 

Can't see the build errors - only see a ```Build cannot be found``` error on shippable. 

Any comments on what else is needed here + can you please let me know the shippable errors. 
I can address the errors + comment  re: autoscale threshold type that @davidopp  had along w/ anything else that needs to be done here.  Thanks.

davidopp commented:
Unfortunately I don't think we're going to be able to give this attention until 1.0.

smarterclayton commented:
@pmorie and I will do the bulk of the review for this, but we probably want
whomever has been involved in autoscaling on the Google side to take a look
as well.

On Jul 4, 2015, at 2:29 AM, David Oppenheimer <notifications@github.com>
wrote:

Unfortunately I don't think we're going to be able to give this attention
until 1.0.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/9612#issuecomment-118468747>
.

ramr committed:
Due to rebase, add new parameter [wide] to the autoscaler printer methods.

ramr committed:
Fix rebase newline gofmt error + add generated swagger spec files.

bgrant0607 commented:
cc @davidopp

smarterclayton commented:
As discussed offline, if there are others who are interested in picking up
the autoscaler implementation and run with it for 1.1 we'd be happy to
assist or hand that off.

On Tue, Jul 28, 2015 at 7:08 AM, Brian Grant <notifications@github.com>
wrote:

> cc @davidopp <https://github.com/davidopp>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9612#issuecomment-125560305>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

ramr committed:
Rebase PR for recent changes - move from using helper to the new storage
interface, regenerate docs and swagger spec and fixup tests for new storage
interface.

jszczepkowski commented:
The updated design of simplified pod autoscaler is described in #12087. We are planning to implement it as a part of 1.1.

ramr commented:
@jszczepkowski  yeah saw that afterwards. Rebased as it was outdated - was on the wait queue for a while now (almost a couple of months). Anyway that's a moot point now ... 

@davidopp  want me to close this out?

ncdc commented:
This is superseded by 12087, right?

@jszczepkowski @ramr

ramr commented:
@ncdc  yes that is correct - finishing up the paperwork here. 
Closing in favor of https://github.com/kubernetes/kubernetes/issues/12087
